numpy/lib/function_base.py

```diff
@@ -3929,21 +3929,35 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        # the linear interpolation below ensure monotonicity with floating
+        # point operations and was proposed in:
+        # https://math.stackexchange.com/a/1798323 by Pedro Gimeno
+        diff_x2_x1 = x2 - x1
+        r_above = x1 + diff_x2_x1 * weights_above
+        r_below = x2 - diff_x2_x1 * weights_below
 
         # ensure axis with q-th is first
-        x1 = np.moveaxis(x1, axis, 0)
-        x2 = np.moveaxis(x2, axis, 0)
+        r_above = np.moveaxis(r_above, axis, 0)
+        r_below = np.moveaxis(r_below, axis, 0)
 
         if zerod:
-            x1 = x1.squeeze(0)
-            x2 = x2.squeeze(0)
+            weights_above = weights_above.squeeze(0)
+            weights_below = weights_below.squeeze(0)
+            r_above = r_above.squeeze(0)
+            r_below = r_below.squeeze(0)
+
+        r = np.where(weights_above < 0.5, r_above, r_below)
```

Comment:
I wonder if we can avoid the intermediate array here when an `out` argument is specified (perhaps via `np.copyto(..., where=...)`, or whether we care.

At the least, we shouldadd a TODO here, perhaps:
```suggestion
        # TODO: use the out argument to `np.where` once we add it (gh-XXXX)
        r = np.where(weights_above < 0.5, r_above, r_below)
```

I think there is an existing issue for an out argument to `where` somewhere.

Reply:
`r_above` (the original) is not needed anymore, so can't we just do `r_above[r_above < 0.5] = r_below`?